File: Day6/Day6.py
def problem1(file: str) -> str:
    """
    Sol to problem 1
    """
    lines = []
    with open(file, "r") as f:
        for line in f:
            lines.append(line.strip())

    total = 0

    # Loop through and compare.
    for i in range(len(lines[0])):
        one = lines[0][i]
        two = lines[0][i + 1]
        three = lines[0][i + 2]
        four = lines[0][i + 3]
        if (
            one == two
            or one == three
            or one == four
            or two == three
            or two == four
            or three == four
        ):
            total += 1
        else:
            return (
                total + 4
            )  # This is needed to compensate for the fact we have to return the end pos and not the start due to characters processed.


def problem2(file: str) -> str:
    """
    Sol to problem2
    """
    lines = []
    with open(file, "r") as f:
        for line in f:
            lines.append(line.strip())

    total = 0

    for i in range(len(lines[0])):
        temp = {}
        for j in range(14):
            if temp.get(lines[0][i + j]):
                temp[lines[0][i + j]] += 1
            else:
                temp[lines[0][i + j]] = 1
        if len(temp) != 14:
            total += 1
        else:
            return (
                total + 14
            )  # Need to add 14 again instead of start, this way we do not have to keep track of characters processed.


print(problem1("data.txt"))
print(problem2("data.txt"))

File: Day6/Day6.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Hashtable;
import java.util.Scanner;

class Day6 {

    public static int problem1(String filename) throws FileNotFoundException
    {
        File f = new File(filename);
        Scanner s = new Scanner(f);
        try
        {
            ArrayList<String> lines = new ArrayList<String>();
            while (s.hasNextLine())
            {
                lines.add(s.nextLine().strip());
            }

            // Total to save total characters processed -4, data to be quicker.
            int total = 0;
            String data = lines.get(0);

            // Looping through values.
            for (int i = 0; i < data.length(); i++) {
                char first = data.charAt(i);
                char second = data.charAt(i + 1);
                char third = data.charAt(i + 2);
                char fourth = data.charAt(i + 3);
                if (first == second || first == third || first == fourth || second == third || second == fourth || third == fourth) {
                    total += 1;
                } else {
                    return total + 4; 
                }
            }
            return 0; // So the compulor stops yelling at me
        }
        finally
        {
            s.close();
        }
    }

    public static int problem2(String filename) throws FileNotFoundException
    {
        File f = new File(filename);
        Scanner s = new Scanner(f);
        try
        {
            ArrayList<String> lines = new ArrayList<String>();
            while (s.hasNextLine())
            {
                lines.add(s.nextLine().strip());
            }

            // Total to save total characters processed -4, data to be quicker.
            int total = 0;
            String data = lines.get(0);

            // Looping through values.
            for (int i = 0; i < data.length(); i++) {
                Hashtable<Character, Integer> temp = new Hashtable<Character, Integer>();
                for (int j = 0; j < 14; j++) {
                    if (temp.get(Character.valueOf(data.charAt(i + j))) != null) {
                        temp.put(Character.valueOf(data.charAt(i + j)), Integer.valueOf(temp.get(Character.valueOf(data.charAt(i + j))) + 1));
                    } else {
                        temp.put(Character.valueOf(data.charAt(i + j)), Integer.valueOf(0));
                    }
                }
                if (temp.size() != 14) {
                    total += 1;
                } else {
                    return total + 14; // Add 14 to ocmpensate for characters processed 
                }
            }
            return 0; // So the compulor stops yelling at me
        }
        finally
        {
            s.close();
        }
    }
    
    public static void main(String[] args) {
        try {
            System.out.println(problem1("data.txt"));
            System.out.println(problem2("data.txt"));
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}